### ISSUE TITLE
can_put_xmp doesn't say why

### ISSUE THREAD

**Issue by johnscancella** [python-xmp-toolkit/python-xmp-toolkit#47]
I have been trying for the better part of a day to try and add a XMP file to a PDF/A file but every time I try `xmpfile.can_put_xmp(xmp)` it returns false. I have no idea why it won't allow me to write it. Is there a way to determine why it is failing easily? I attached the files I am using (had to rename the xmp file as github wouldn't allow me to upload it).

The code I am running:
```
#!/usr/bin/env python3

from libxmp import XMPFiles

xmpfile = XMPFiles( file_path="output1.pdf", open_forupdate=True )
xmpdata = XMPFiles( file_path="004004.modified.xmp" )
xmp = xmpdata.get_xmp()

if xmpfile.can_put_xmp(xmp):
  xmpfile.put_xmp(xmp)
  xmpfile.close_file()
else:
  print("can't put new xmp metadata!")

[output1.pdf](https://github.com/python-xmp-toolkit/python-xmp-toolkit/files/114456/output1.pdf)
[004004.modified.xmp.txt](https://github.com/python-xmp-toolkit/python-xmp-toolkit/files/114460/004004.modified.xmp.txt)




```

**Comment by quintusdias** [comment]
Hey John, sorry for the late answer, which is that you can't use the exempi
library to inject XMP into a PDF that does not already have an XMP packet.
Some background on the issue here, https://forums.adobe.com/thread/708717 .

On Tue, Feb 2, 2016 at 1:37 PM, John Scancella <notifications@github.com>
wrote:

> I have been trying for the better part of a day to try and add a XMP file
> to a PDF/A file but every time I try xmpfile.can_put_xmp(xmp) it returns
> false. I have no idea why it won't allow me to write it. Is there a way to
> determine why it is failing easily? I attached the files I am using (had to
> rename the xmp file as github wouldn't allow me to upload it).
>
> The code I am running:
>
> #!/usr/bin/env python3
>
> from libxmp import XMPFiles
>
> xmpfile = XMPFiles( file_path="output1.pdf", open_forupdate=True )
> xmpdata = XMPFiles( file_path="004004.modified.xmp" )
> xmp = xmpdata.get_xmp()
>
> if xmpfile.can_put_xmp(xmp):
>   xmpfile.put_xmp(xmp)
>   xmpfile.close_file()
> else:
>   print("can't put new xmp metadata!")
>
> [output1.pdf](https://github.com/python-xmp-toolkit/python-xmp-toolkit/files/114456/output1.pdf)
> [004004.modified.xmp.txt](https://github.com/python-xmp-toolkit/python-xmp-toolkit/files/114460/004004.modified.xmp.txt)
>
>
>
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/python-xmp-toolkit/python-xmp-toolkit/issues/47>.
>



-- 
John Evans

**Comment by johnscancella** [comment]
@quintusdias but I am not trying to inject XMP into a PDF that does not already have an XMP packet. I am trying to MERGE a XMP file into a PDF that already has XMP.

**Comment by quintusdias** [comment]
So after debugging the exempi library, it appears that in the case of PDF, not only does an XMP packet have to exist, but the packet that you want to merge back in has to be smaller than the pre-existing packet?  Line 1199 of ``exempi-2.2.2/source/XMPCore/XMPMeta-Serialize.cpp`` is where that check occurs and where your case fails (which is in the Adobe SDK being wrapped by exempi).  By manually increasing the padding at the bottom of the XMP packet in the PDF file, it can be made to work, but that's obviously not a great solution.

It's unfortunate, but about the only thing I can think to do is to somehow get a better diagnostic.

**Comment by johnscancella** [comment]
@quintusdias Ok that makes sense. If I wanted to do the "not great solution" of manually increasing the padding at the bottom, how would I go about doing that?

**Comment by quintusdias** [comment]
The way I did it was to put blank space between the final </x:xmpmeta>  and <?xpacket end="w"?> .  The XMP specification refers to this as the "padding" and recommends that it be 2-4 KB, seemingly for this very purpose.   If you end up having to go to lengths like that, you might as well bypass the python xml toolkit entirely and manually excise the old XMP and stitch in the new.

**Comment by johnscancella** [comment]
Yeah that is what I was afraid of. Ok, then can I change this issue so that the underlaying exempi code error messages come back instead of just a boolean for can put?
For example, the exempi code you referenced above, could xmp-toolkit supply an optional error message `Can't fit into specified packet size` ?

**Comment by johnscancella** [comment]
Also I submitted a bug report to exempi for this underlying issue
https://bugs.freedesktop.org/show_bug.cgi?id=94065

**Comment by hfiguiere** [comment]
Given Adobe answer is https://forums.adobe.com/message/3096497#3096497

"This is because the toolkit does not have a smart file handler for PDF files but just uses dump packet scanning when interacting with XMP in that file format.

The PDF library should be able to inject XMP into PDF files." 

It is unlikely to get fixed quick. Another case where they didn't want to disrupt their business (selling the PDF library).

see https://forums.adobe.com/message/3099970#3099970
"the request for a smart PDF handler exists for a long time already but so far demand was unfortunately never high enough that it would make it to the top of the priority list."

What I would recommend for your case would be to special case PDF and use a PDF writing library to write the XMP packet to PDF.

**Comment by hfiguiere** [comment]
(I'm the maintainer of Exempi, BTW)

**Comment by hfiguiere** [comment]
But it seems that Exempi is missing some APIs for a better control on the packet serialization and size that the SDK has. I could roll a 2.3 release with these. (you'd need to update the python bindings too).

**Comment by quintusdias** [comment]
Hey Hubert, pleased to make acquaintance with you.

A 2.3 release would be great, but I agree that making a special case for PDF might work best for now.

**Comment by johnscancella** [comment]
@quintusdias looks like 2.3 was released with the fix https://bugs.freedesktop.org/show_bug.cgi?id=94065
